### PR TITLE
New feature #17477: Make it possible to use answer option text in expression script.

### PR DESCRIPTION
**Pull request by DenisChenu**
Dev: plugin added

- [x] Add test

### CODE REVIEW

File: application/config/third_party.php [hunk 1]
```diff
@@ -416,8 +416,9 @@
             'leaflet.css'
         ),
     ),
+    /* For geoname search autocomplete without jquery */
     'devbridge-autocomplete' => array(
-        'basePath' => 'third_party.devbridge-autocomplete.dist', /* For geoname search autocomplete without jquery */
```
Comment (gabrieljenik) [inline application/config/third_party.php:420]: Was this intended to be removed on this PR?

Comment (Shnoulle) [reply application/config/third_party.php:420]: No … but it was not removed, just move comments. I fix it :)

**Review by gabrieljenik — APPROVED** [review]

File: application/core/plugins/expressionAnswerOptions/AnswerOptionsFunctions.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+use Question;
+use Answer;
+use Permission;
+use CHtml;
+
+class AnswerOptionsFunctions
+{
+    /**
+     * Return the answer text related to a question
+     * @param integer|string $qid : id or code of question, get parent question if needed
```
Comment (olleharstedt) [inline application/core/plugins/expressionAnswerOptions/AnswerOptionsFunctions.php:19]: Param has different name than actual argument

File: application/core/plugins/expressionAnswerOptions/AnswerOptionsFunctions.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+use Permission;
+use CHtml;
+
+class AnswerOptionsFunctions
+{
+    /**
+     * Return the answer text related to a question
+     * @param integer|string $qid : id or code of question, get parent question if needed 
+     * @param string $code : code of the answer text to return
+     * @return null|string
```
Comment (olleharstedt) [inline application/core/plugins/expressionAnswerOptions/AnswerOptionsFunctions.php:21]: $scale is not documented

File: application/core/plugins/expressionAnswerOptions/expressionAnswerOptions.php [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+class expressionAnswerOptions extends PluginBase
```
Comment (olleharstedt) [inline application/core/plugins/expressionAnswerOptions/expressionAnswerOptions.php:23]: PSR-12 require capital case class name, like ExpressionAnswerOptions

**Review by olleharstedt — CHANGES_REQUESTED** [review]

File: composer.json [hunk 5]
```diff
@@ -63,6 +63,8 @@
             "./third_party/bin/phpcs --standard=tests/rulesets/phpcs_ruleset.xml application/models/ArchivedTableSettings.php",
             "./third_party/bin/phpcs --standard=tests/rulesets/phpcs_ruleset.xml application/libraries/ExtensionInstaller",
             "./third_party/bin/phpcs --standard=tests/rulesets/phpcs_ruleset.xml application/helpers/update/updatedb_helper.php",
+            "./third_party/bin/phpcs --standard=tests/rulesets/phpcs_ruleset.xml application/core/plugins/ExpressionAnswerOptions/ExpressionAnswerOptions.php",
+            "./third_party/bin/phpcs --standard=tests/rulesets/phpcs_ruleset.xml application/core/plugins/ExpressionAnswerOptions/AnswerOptionsFunctions.php",
```
Comment (olleharstedt) [inline composer.json:67]: Can you try phpmd_ruleset_strict.xml, please? It sets a limit on function length of 60 LOC.

File: tests/functional/frontend/AnswerOptionsFunctionsPluginTest.php [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+    /**
+     * Check if function return good value in loaded survey
+     **/
+    public function testGetAnswerOptionText()
+    {
+        $urlMan = \Yii::app()->urlManager;
+        $urlMan->setBaseUrl('http://' . self::$domain . '/index.php');
+
+        $testToDo = array(
```
Comment (olleharstedt) [inline tests/functional/frontend/AnswerOptionsFunctionsPluginTest.php:36]: Tiny details, but personally I prefer the `[]` notation above the older `array()`.

Comment (DenisChenu) [reply tests/functional/frontend/AnswerOptionsFunctionsPluginTest.php:36]: Personally : i prefer array( … )

A ruleset for this ?

File: composer.json (hunk 5 above)
Comment (DenisChenu) [reply composer.json:67]: You know you can update file directly ?

File: tests/functional/frontend/AnswerOptionsFunctionsPluginTest.php (hunk 6 above)
Comment (olleharstedt) [reply tests/functional/frontend/AnswerOptionsFunctionsPluginTest.php:36]: Any reason you prefer the old notation? :)

Have to check for conventions...

Comment (DenisChenu) [reply tests/functional/frontend/AnswerOptionsFunctionsPluginTest.php:36]: I'm old too ;) 

but : 
array() : «OK, it's an array»

boolval($toto) : ok i get a boolean value

etc  …

Comment (olleharstedt) [reply tests/functional/frontend/AnswerOptionsFunctionsPluginTest.php:36]: https://www.reddit.com/r/PHP/comments/3qekc8/cs_array_or/

File: application/libraries/PluginManager/PluginBase.php [hunk 7]
```diff
@@ -251,13 +251,12 @@ protected function registerSetting($name, $options = array('type' => 'string'))
     }
 
     /**
-     *
      * @param array<string, mixed> $settings
      * @return void
      */
-    public function saveSettings($settings)
+    public function saveSettings($aSettings)
```
Comment (olleharstedt) [inline application/libraries/PluginManager/PluginBase.php:257]: Hm why this change? Hungarian notation should be used less, not more :stuck_out_tongue:

File: composer.json (hunk 5 above)
Comment (olleharstedt) [reply composer.json:67]: Hm, gonna try

File: tests/functional/frontend/AnswerOptionsFunctionsPluginTest.php (hunk 6 above)
Comment (DenisChenu) [reply tests/functional/frontend/AnswerOptionsFunctionsPluginTest.php:36]: https://www.reddit.com/r/PHP/comments/3qekc8/comment/cwel0ay/?utm_source=share&utm_medium=web2x&context=3 

But : need to be set in a ruleset.

File: application/libraries/PluginManager/PluginBase.php (hunk 7 above)
Comment (DenisChenu) [reply application/libraries/PluginManager/PluginBase.php:257]: Because : it must be set here because it was set in LimeSurvey\PluginManager\iPlugin::saveSettings

See https://github.com/LimeSurvey/LimeSurvey/runs/4444187513?check_suite_focus=true#step:9:34

>  ERROR: ParamNameMismatch - application/core/plugins/ExpressionAnswerOptions/ExpressionAnswerOptions.php:93:34 - Argument 1 of ExpressionAnswerOptions::saveSettings has wrong name $settings, expecting $aSettings as defined by LimeSurvey\PluginManager\iPlugin::saveSettings (see https://psalm.dev/230)
>    public function saveSettings($settings)

Comment (olleharstedt) [reply application/libraries/PluginManager/PluginBase.php:257]: lol ok, thanks

File: tests/functional/frontend/AnswerOptionsFunctionsPluginTest.php (hunk 6 above)
Comment (olleharstedt) [reply tests/functional/frontend/AnswerOptionsFunctionsPluginTest.php:36]: Yep, will add a task for it

**Comment by DenisChenu** [comment]
Argl … i think rename aSettings to $settings is a better solution ;) I do it directly

File: application/libraries/PluginManager/PluginBase.php (hunk 7 above)
Comment (DenisChenu) [reply application/libraries/PluginManager/PluginBase.php:257]: Rename LimeSurvey\PluginManager\iPlugin::saveSettings parameters seem better … i fix it

File: tests/functional/frontend/AnswerOptionsFunctionsPluginTest.php (hunk 6 above)
Comment (DenisChenu) [reply tests/functional/frontend/AnswerOptionsFunctionsPluginTest.php:36]: Minimum : https://manual.limesurvey.org/Code_quality_guide , if it was in manual : i respect :)
